flink-runtime/src/main/java/org/apache/flink/runtime/controlplane/dispatcher/StreamManagerDispatcher.java

```diff
@@ -55,20 +63,78 @@ private void startDispatcherService() {
 
 	}
 
-	private CompletableFuture<StreamManagerRunner> startStreamManagerRunner(JobGraph jobGraph) {
-		final RpcService rpcService = getRpcService();
+	public CompletableFuture<Acknowledge> submitJob(JobGraph jobGraph, Time timeout) {
+		// SMD refers to StreamManagerDispatcher
+		log.info("[SMD] Received JobGraph submission {} ({})", jobGraph.getJobID(), jobGraph.getName());
+		// TODO: deal with duplicates and partial resource
+		return internalSubmitJob(jobGraph);
+	}
+
+	private CompletableFuture<Acknowledge> internalSubmitJob(JobGraph jobGraph) {
+		log.info("[SMD] Submitting job {} ({})", jobGraph.getJobID(), jobGraph.getName());
+		// TODO: waitForTerminatingStreamManager
+		final CompletableFuture<Acknowledge> submitDirectly = persistAndRunJob(jobGraph)
+				.thenApply(ignored -> Acknowledge.get());
+		return submitDirectly.handleAsync((acknowledge, throwable) -> {
+			if (throwable != null) {
+				//TODO: cleanup data
+
+				final Throwable strippedThrowable = ExceptionUtils.stripCompletionException(throwable);
+				log.error("Failed to submit job {}.", jobGraph.getJobID(), strippedThrowable);
+				throw new CompletionException(
+						new JobSubmissionException(jobGraph.getJobID(), "Failed to submit job.", strippedThrowable));
+			} else {
+				return acknowledge;
+			}
+		}, getRpcService().getExecutor());
+	}
+
+	private CompletableFuture<Void> persistAndRunJob(JobGraph jobGraph) {
+		// TODO: jobGraphWriter
+		final CompletableFuture<Void> runSMFuture = runStream(jobGraph);
+		return runSMFuture.whenComplete(BiConsumerWithException.unchecked((Object ignored, Throwable throwable) -> {
+			if (throwable != null) {
+				final Throwable strippedThrowable = ExceptionUtils.stripCompletionException(throwable);
+
+				log.error("Failed to run stream manager.", strippedThrowable);
+				// jobGraphWriter.removeJobGraph(jobGraph.getJobID());
+			}
+		}));
+	}
 
+	private CompletableFuture<Void> runStream(JobGraph jobGraph) {
```

Comment:
这个方法是干什么的？如果是run stream manager 建议方法名字改为runStreamManager()
```suggestion
	private CompletableFuture<Void> runStreamManager(JobGraph jobGraph) {
```

Reply:
对的，Dispatcher里有对应的`runJob()` method。以后我们可能需要的也是`runJob()`，在里面创建StreamManager。

提议不错，立马修改。